### PR TITLE
Added check for invalid quaternions.

### PR DESCRIPTION
**Pull request by StefanFabian**
Added checks for quaternions before passing them to OGRE to fix crashes due to unnormalized quaternions as described in #1142.

### CODE REVIEW

**Comment by wjwwood** [comment]
Thanks for pr, one of us will get to it asap.

**Comment by VictorLamoine** [comment]
@wjwwood can this PR be merged? It's quite annoying to crash RViz because of this :)

**Comment by StefanFabian** [comment]
After reading #1113, I changed the copy right statement.

File: src/rviz/validate_quaternions.h [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+#include <tf/LinearMath/Quaternion.h>
+
+#include <boost/array.hpp>
+
+namespace rviz
+{
+
+inline bool validateQuaternions( float w, float x, float y, float z )
+{
+  return std::abs( w * w + x * x + y * y + z * z - 1.0f ) < 10e-5f;
```
Comment (dhood) [inline src/rviz/validate_quaternions.h]: Was there a particular reason you switched to using `10e-5`? If not, in tf2 they found the tolerance had to be relaxed it to `10e-3` (https://github.com/ros/geometry2/commit/634fd0cca9d8eb54a2040f5f7f837c9f3156504a), so please stick with `10e-3`

Comment (StefanFabian) [reply src/rviz/validate_quaternions.h]: Not really, I just felt like 10e-3 was a bit too relaxed for my preferences but if empirical evidence proves me wrong, I'll change it back to 10e-3.

**Comment by dhood** [comment]
@ros-pull-request-builder retest this please

**Comment by pbeeson** [comment]
I'm very happy that this might fix problems with IMs crashing RViz that wev'e seen.  However, is there a reason why the patch couldn't just normalize the quaternion in place?

**Comment by StefanFabian** [comment]
Good question.
I'd say there is a reason which is that this would just hide the symptoms of possibly flawed code.
However, just displaying a warning and normalizing would probably be more in line with the behavior before the crashes.

**Comment by pbeeson** [comment]
Yes.  It took me over an hour after upgrading my home desktop this morning to figure out why my Interactive MArker's weren't showing up.  It traced back to a bad Quaternion in the library (not written by me).  The fix was trivial, but if you had enforced normalization here instead of checking, my code would have worked as "normal".

If you took this to an extreme, you could just always normalize no matter what.  Then you wouldn't have to use arbitrary thresholds for the magnitude like 10e-3 (unless that's a threshold that came from OGRE).  Of course the issue here would be a minor slowdown to do 4 extra divides, but that's nothing really.

**Comment by davetcoleman** [comment]
This change broke my workflow today too. Note @rhaschke has proposed a fix: https://github.com/ros-visualization/rviz/pull/1179

**Comment by pbeeson** [comment]
@wjwwood I agree with @davetcoleman #1179 looks much more appropriate.  If it’s zero then error otherwise just normalize.

**Comment by pbeeson** [comment]
I guess it was actually @dhood that approved and merged the original PR.

**Comment by dhood** [comment]
I didn't expect this change to impact so many users since quaternions really ought to be normalised (TF will reject them and roswtf will report them). From the number of responses here I can appreciate that it's simply not a reasonable expectation to make. I agree with @rhaschke's note in https://github.com/ros-visualization/rviz/issues/1178#issuecomment-354488722 that there should have been a smoother transition phase rather than rejecting them, which I will look into ASAP

**Comment by StefanFabian** [comment]
First, I want to apologize that my PR caused this many issues.
If I may still have a word on solving this issue, I'd propose to also allow zeros and just set them to the identity quaternion before this leads to the next issue.
I would, however, not remove the warning because an unnormalized quaternion can indicate that there's an error in the code. It may just be a missing normalization but it could also be a more serious issue.
Hence, I would suggest to replace all returns in this commit by a call to a normalization function which sets all null quaternions to identity quaternions and simply normalizes the rest.

**Comment by rhaschke** [comment]
@StefanFabian This is more or less what I proposed in #1179 ;-)

**Pull request merged**